File: pset6/sentimental-credit/credit.py
from cs50 import get_string


def main():
    card = get_string("Number: ")
    if chksum(card) == 0:
        if (len(card) == 13 or len(card) == 16) and card[0] == "4":
            print("VISA")
        elif len(card) == 15 and int(card[0:2]) in (34, 37):
            print("AMEX")
        elif len(card) == 16 and int(card[0:2]) in range(51, 56):
            print("MASTERCARD")
        else:
            print("INVALID")
    else:
        print("INVALID")


def chksum(card):
    chksum = 0
    count = 0
    for digit in reversed(card):
        # for every other digit starting from last
        if count % 2 == 0:
            chksum += int(digit)
            count += 1
        # if the product is 2 digits then separate them and sum
        elif int(digit) * 2 > 9:
            chksum += int(digit) * 2 // 10 + int(digit) * 2 % 10
            count += 1
        # for every other digit starting from second to last
        else:
            chksum += int(digit) * 2
            count += 1
    chksum %= 10
    return chksum


if __name__ == "__main__":
    main()

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int chksumcalc(long);

int main(void)
{
    // Prompt user for input
    long ccnum = get_long("Number: ");


    // Evaluate chksum
    int chksum = chksumcalc(ccnum);
    if (chksum == 0)
    {
        // evaluate for 13 digits for visa
        if (ccnum < 10000000000000)
        {
            if (ccnum / 1000000000000 == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        // evaluate for 15 digits for amex
        else if (ccnum < 1000000000000000)
        {
            if (ccnum / 10000000000000 == 34 || ccnum / 10000000000000 == 37)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        // evaluate for 16 digits
        else if (ccnum < 10000000000000000)
        {
            // evaluate first digits 51-55 for mastercard
            if (ccnum / 100000000000000 >= 51 && ccnum / 100000000000000 <= 55)
            {
                printf("MASTERCARD\n");
            }
            // evaluate first digit 4 for visa
            else if (ccnum / 1000000000000000 == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

// chksum calculator function
int chksumcalc(long ccnum)
{
    int chksum = 0;
    int seconddigitsum = 0;
    for (int digitcount = 0; digitcount < 8; digitcount++)
    {
        // add last digit to chksum
        chksum = ccnum % 10 + chksum;
        // second digit operations
        seconddigitsum = ccnum / 10 % 10 * 2;
        if (seconddigitsum > 9)
        {
            // extrapolate digits and add together
            seconddigitsum = (seconddigitsum % 10) + (seconddigitsum / 10 % 10);
            // add extrapolated digits' sum to chksum
            chksum = seconddigitsum + chksum;
        }
        else
        {
            // otherwise add second to last digit *2 to chksum
            chksum = seconddigitsum + chksum;
        }
        //drop last 2 digits to move on
        ccnum = ccnum / 100;
    }
    // obtain last digit of chksum
    chksum = chksum % 10;

    return chksum;
}

